File: 1601-maximum-number-of-achievable-transfer-requests/1601-maximum-number-of-achievable-transfer-requests.java
class Solution {
    int ans = 0;
    public void solve(int i, int[][] requests, int numberOfRequestProcessed, int[] requestProcessed){
        
        if(i >= requests.length){
            
            for(int r : requestProcessed)
                if(r != 0)
                    return;
            
            ans = Math.max(ans, numberOfRequestProcessed);
            return;
        }
        
        solve(i+1, requests, numberOfRequestProcessed, requestProcessed);
        
        requestProcessed[requests[i][0]]--;
        requestProcessed[requests[i][1]]++;
        
        solve(i+1, requests, 1+numberOfRequestProcessed, requestProcessed);
        
        requestProcessed[requests[i][0]]++;
        requestProcessed[requests[i][1]]--;
    }
    
    public int maximumRequests(int n, int[][] requests) {
        
        int[] requestProcessed = new int[n];
        
        solve(0, requests, 0, requestProcessed);
        
        return ans;
    }
}

File: 1601-maximum-number-of-achievable-transfer-requests/1601-maximum-number-of-achievable-transfer-requests.cpp
class Solution {
public:
    int ans=0;
    void solve(int i, vector<vector<int>>& requests, int numberOfRequestsProcessed, vector<int>& requestProcessed){
        
        if(i == requests.size()){
            
            for(auto r: requestProcessed){
                if(r != 0)
                    return;
            }
            
            ans = max(ans, numberOfRequestsProcessed);
            return;
        }
        
        // not processing the request
        solve(i+1, requests, numberOfRequestsProcessed, requestProcessed);
        
        // process this request
        requestProcessed[requests[i][0]]--;
        requestProcessed[requests[i][1]]++;
        
        solve(i+1, requests, 1+numberOfRequestsProcessed, requestProcessed);
        
        requestProcessed[requests[i][0]]++;
        requestProcessed[requests[i][1]]--;
    }
    
    int maximumRequests(int n, vector<vector<int>>& requests) {
        
        vector<int> requestProcessed(n, 0);
        solve(0, requests, 0, requestProcessed);
        
        return ans;
    }
};